### ISSUE TITLE
Reconcile multiple language match by suffix length

### ISSUE THREAD

**Issue by jianghoy** [zed-industries/zed#8408]
### Check for existing issues

- [X] Completed

### Describe the feature

Inspired by https://github.com/zed-industries/zed/issues/7096#issuecomment-1948414691.

# The problem
Today, when multiple `LanguageMatcher`'s path_suffix match a file,  one random language will be chosen (see [language_registry.rs get_or_load_language function](https://github.com/zed-industries/zed/blob/ffdda588b41f7d9d270ffe76cab116f828ad545e/crates/language/src/language_registry.rs#L347-L358)). While this simple setting works well for most of the cases, it's not robust enough. E.g. there's no way to distinguish `.php` or `.blade.php`.

# Proposed fix
To make it simple, I propose changing [the callback function](https://github.com/zed-industries/zed/blob/ffdda588b41f7d9d270ffe76cab116f828ad545e/crates/language/src/language_registry.rs#L342) so that it returns a tuple of boolean + max matched suffix length among all the suffixes. When there're multiple matches, the language with the longest suffix wins.

**Comment by maxbrunsfeld** [comment]
I agree this would be a clear improvement. In PR #11697, @claytonrcarter did a bit of work trying to implement it. If anyone has the time to finish up that work, that would be great.

**Comment by claytonrcarter** [comment]
> In PR [#11697](https://github.com/zed-industries/zed/pull/11697), [@claytonrcarter](https://github.com/claytonrcarter) did a bit of work trying to implement it. If anyone has the time to finish up that work, that would be great.

Thanks for the shout out, Max. I obviously haven't been able to get back to #11697, but I have at least been keeping the code rebased and building. If anyone wants to pick it up, feel free to use the code at https://github.com/claytonrcarter/zed/tree/extension-length as a starting point, if you'd like. As I recall, when I left it, I had resolved most of the review comments, but glob support had recently been added and I got stumped on how globs should factor in to "length of matched suffix length". I believe that the code in the above branch just counts the string length of the glob as if it were a static suffix (ie `*js` = 3 and `js` = 2) ... that is probably fine for most cases, but I just haven't had time to get back to double check and to polish it again for a PR.
